File: goopang-java/src/main/java/com/example/goopangjava/resolver/ProductResolver.java
package com.example.goopangjava.resolver;

import com.example.goopangjava.entity.product.Product;
import com.example.goopangjava.input.AddProductInput;
import com.example.goopangjava.service.ProductService;
import org.apache.coyote.BadRequestException;
import org.springframework.graphql.data.method.annotation.Argument;
import org.springframework.graphql.data.method.annotation.MutationMapping;
import org.springframework.graphql.data.method.annotation.QueryMapping;
import org.springframework.graphql.data.method.annotation.SubscriptionMapping;
import org.springframework.stereotype.Controller;
import reactor.core.publisher.Flux;

import java.util.List;

@Controller
public class ProductResolver {
    private final ProductService productService;

    public ProductResolver(ProductService productService) {
        this.productService = productService;
    }

    @QueryMapping
    public List<Product> getProducts() {
        return productService.getProducts();
    }

    @MutationMapping
    public Product addProduct(@Argument AddProductInput addProductInput) throws BadRequestException {
        return productService.addProduct(addProductInput);
    }

    @SubscriptionMapping
    public Flux<Product> newProduct(@Argument String productName) {
        return productService.messageFlux(productName);
    }
}

File: goopang-kotlin/src/main/kotlin/com/example/goopang/resolver/ProductResolver.kt
package com.example.goopang.resolver

import com.example.goopang.entity.product.Product
import com.example.goopang.input.AddProductInput
import com.example.goopang.service.ProductService
import org.springframework.graphql.data.method.annotation.Argument
import org.springframework.graphql.data.method.annotation.MutationMapping
import org.springframework.graphql.data.method.annotation.QueryMapping
import org.springframework.graphql.data.method.annotation.SubscriptionMapping
import org.springframework.stereotype.Controller
import reactor.core.publisher.Flux

@Controller
class ProductResolver(
    private val productService: ProductService
) {

    @QueryMapping
    fun getProducts(): ArrayList<Product> {
        return productService.getProducts()
    }

    @MutationMapping
    fun addProduct(
        @Argument addProductInput: AddProductInput
    ): Product {
        return productService.addProduct(addProductInput)
    }

    @SubscriptionMapping
    fun newProduct(
        @Argument productName: String?
    ): Flux<Product> {
        return productService.messageFlux(productName)
    }
}